socs/agents/lakeshore240/agent.py

```diff
@@ -108,202 +148,211 @@ def set_values(self, session, params=None):
+
+            # For session.data
+            field_dict[chan_string] = {"T": temp_reading, "V": sensor_reading}
+
+        data = {
+            "timestamp": current_time,
+            "block_name": "temps",
+            "data": data_dict,
+        }
 
-        f_sample = params.get('sampling_frequency')
-        # If f_sample is None, use value passed to Agent init
-        if f_sample is None:
-            f_sample = self.f_sample
-
-        sleep_time = 1 / f_sample - 0.01
-
-        with self.lock.acquire_timeout(0, job='acq') as acquired:
-            if not acquired:
-                self.log.warn("Could not start acq because {} is already running"
-                              .format(self.lock.job))
-                return False, "Could not acquire lock."
-
-            self.take_data = True
-
-            session.data = {"fields": {}}
-
-            while self.take_data:
-                current_time = time.time()
-                data = {
-                    'timestamp': current_time,
-                    'block_name': 'temps',
-                    'data': {}
-                }
-
-                for chan in self.module.channels:
-                    # Read sensor on channel
-                    chan_string = "Channel_{}".format(chan.channel_num)
-                    temp_reading = chan.get_reading(unit='K')
-                    sensor_reading = chan.get_reading(unit='S')
-
-                    # For data feed
-                    data['data'][chan_string + '_T'] = temp_reading
-                    data['data'][chan_string + '_V'] = sensor_reading
-
-                    # For session.data
-                    field_dict = {chan_string: {"T": temp_reading, "V": sensor_reading}}
-                    session.data['fields'].update(field_dict)
-
-                self.agent.publish_to_feed('temperatures', data)
-
-                session.data.update({'timestamp': current_time})
-
-                time.sleep(sleep_time)
-
-            self.agent.feeds['temperatures'].flush_buffer()
-
-        return True, 'Acquisition exited cleanly.'
+        session.data["fields"] = field_dict
+        self.agent.publish_to_feed("temperatures", data)
+        session.data["timestamp"] = current_time
+        return data
 
-    def _stop_acq(self, session, params=None):
+    def _process_actions(self, module: Module) -> None:
+        """
+        Processes queued actions using the provided Lakeshore Module.
         """
-        Stops acq process.
+        while not self.action_queue.empty():
+            action = self.action_queue.get()
+            try:
+                log.info(f"Running action {action}")
+                res = action.process(module)
+                action.deferred.callback(res)
+            except Exception as e:  # pylint: disable=broad-except
+                log.error(f"Error processing action: {action}")
+                action.deferred.errback(e)
+        return None
+
+    def main(
```

Comment:
That's a good question, I think there isn't a clear answer how multiple processes should work. In general, multiple processes like that in general kind of breaks the hardware "single point of access" I'm going for here. I think it would depend on a case to case basis. Like you say, for the custom_pid function would probably be an argument you could pass into the main function to swap out the behavior.